### PR TITLE
Opt, Dev mode for raven_framework

### LINKED ISSUE #408: Speedup through asserts
--------
Issue Description
--------
##### What did you expect to see happen?
API checking should not get in the way of fast performance.

##### What did you see instead?
Some Python `if` checks dealing with APIs are impossible for the user to hit, and
when called millions of times per run, can slow down RAVEN calculations.

##### Do you have a suggested fix for the development team?
Python has `assert` functionality that provides API checks; however, if python is run with the `-O` option, these assertions are skipped.  We could make use of this to have a `dev` mode and an `opt` mode.

----------------
For Change Control Board: Issue Review
----------------
This review should occur before any development is performed as a response to this issue.
- [x] 1. Is it tagged with a type: defect or improvement?
- [x] 2. Is it tagged with a priority: critical, normal or minor?
- [x] 3. If it will impact requirements or requirements tests, is it tagged with requirements?
- [x] 4. If it is a defect, can it cause wrong results for users? If so an email needs to be sent to the users.
- [x] 5. Is a rationale provided? (Such as explaining why the improvement is needed or why current code is wrong.)

-------
For Change Control Board: Issue Closure
-------
This review should occur when the issue is imminently going to be closed.
- [x] 1. If the issue is a defect, is the defect fixed? __Not a defect__
- [x] 2. If the issue is a defect, is the defect tested for in the regression test system?  (If not explain why not.) __N/A__
- [x] 3. If the issue can impact users, has an email to the users group been written (the email should specify if the defect impacts stable or master)?  __email was proposed in PR #409__ 
- [x] 4. If the issue is a defect, does it impact the latest stable branch? If yes, is there any issue tagged with stable (create if needed)? __Not a defect__
- [x] 5. If the issue is being closed without a merge request, has an explanation of why it is being closed been provided? __PR #409__

Issue close checklist is reviewed by Congjian Wang @wangcj05 on Oct. 26, 2017.

### PR DESCRIPTION
**Pull request by PaulTalbot-INL**
--------
Pull Request Description
--------
Closes #408 

#### What are the significant changes in functionality due to this change request?
By default, sets Python to run in `-O` optimized mode, which strips out assertions and creates `.pyo` files.  This allows developers to use `assert` statements within Python code.  Suggestions of when this should be used are detailed [in the wiki](https://github.com/idaholab/raven/wiki/Python-Optimization).

Also adds a "help" option to `raven_framework` that details command-line options.

----------------
For Change Control Board: Change Request Review
----------------
The following review must be completed by an authorized member of the Change Control Board.
- [x] 1. Review all computer code.
- [x] 2. If any changes occur to the input syntax, there must be an accompanying change to the user manual and xsd schema. If the input syntax change deprecates existing input files, a conversion script needs to be added (see Conversion Scripts).
- [x] 3. Make sure the Python code and commenting standards are respected (camelBack, etc.) - See on the [wiki](https://github.com/idaholab/raven/wiki/RAVEN-Code-Standards#python) for details.
- [x] 4. Automated Tests should pass, including run_tests, pylint, manual building and xsd tests. If there are changes to Simulation.py or JobHandler.py the qsub tests must pass. __All Pass__
- [x] 5. If significant functionality is added, there must  be tests added to check this. Tests should cover all possible options.  Multiple short tests are preferred over one large test. If new development on the internal JobHandler parallel system is performed, a cluster test must be added setting, in <RunInfo> XML block, the node ```<internalParallel>``` to True. __N/A__
- [x] 6. If the change modifies or adds a requirement or a requirement based test case, the Change Control Board's Chair or designee also needs to approve the change.  The requirements and the requirements test shall be in sync.
- [x] 7. The merge request must reference an issue.  If the issue is closed, the issue close checklist shall be done. #408
- [x] 8. If an analytic test is changed/added is the the analytic documentation updated/added? __N/A__

Checklists were reviewed by Congjian Wang @wangcj05 on Oct. 26, 2017

### CODE REVIEW

File: raven_framework [hunk 1]
```diff
@@ -11,5 +11,54 @@ SCRIPT_DIR=`(cd $SCRIPT_DIRNAME; pwd)`
...
+  echo '    Do not attempt to run the model executable, but read from existing output files.'
+  echo ''
+  echo '  interactive'
+  echo '    Run RAVEN in GUI mode if possible.'
+  echo ''
+  echo '  interactiveCheck'
+  echo '    Run tests on interactive GUI.'
+  echo ''
+elif [ "$MODE" = 'opt' ]; then
+  python -O $SCRIPT_DIR/framework/Driver.py "$ARGS"
```
Comment (wangcj05) [inline raven_framework]: I think it is better to print the description before we run. Something like "echo 'Runs RAVEN in 'OPT' mode'"

File: raven_framework [hunk 2]
```diff
@@ -11,5 +11,54 @@ SCRIPT_DIR=`(cd $SCRIPT_DIRNAME; pwd)`
...
+else
+  python $SCRIPT_DIR/framework/Driver.py "$ARGS"
```
Comment (wangcj05) [inline raven_framework]: Same as previous comments.

File: raven_framework [hunk 3]
```diff
@@ -11,5 +11,54 @@ SCRIPT_DIR=`(cd $SCRIPT_DIRNAME; pwd)`
...
+    # display "help" print
+    MODE='help'
+    continue
+  fi
+  # otherwise, pass through arguments
+  ARGS[${#ARGS[@]}]="$var"
+done
+
+# run Driver.py based on the mode chosen
+if [ "$MODE" = 'help' ]; then
```
Comment (wangcj05) [inline raven_framework]: I think we should set $MODE = 'help' if $@ is empty as well

File: raven_framework [hunk 4]
```diff
@@ -11,5 +11,54 @@ SCRIPT_DIR=`(cd $SCRIPT_DIRNAME; pwd)`
...
+# set up run keywords
+# "ARGS" stores command line arguments not treated in this file (passed through)
+declare -a ARGS
+# by default run in "opt" mode
+MODE='opt'
+# loop through arguments and look for those that are handled before Python is called
+for var in "$@"; do
+  if [ "$var" = '-D' ]; then
+    # run in development mode (assertions not stripped)
+    MODE='dev'
```
Comment (wangcj05) [inline raven_framework]: MODE is set to 'dev' but not referenced in the later code.

Comment (PaulTalbot-INL) [reply raven_framework]: I turned the "mode" check into a "case" statement and exit on a catch-all, so dev is explicitly used.

File: raven_framework (hunk 3 above)
Comment (PaulTalbot-INL) [reply raven_framework]: Done.

File: raven_framework (hunk 1 above)
Comment (PaulTalbot-INL) [reply raven_framework]: Done.

File: raven_framework (hunk 2 above)
Comment (PaulTalbot-INL) [reply raven_framework]: Done.

File: raven_framework [hunk 5]
```diff
@@ -31,34 +31,49 @@ for var in "$@"; do
...
+    echo ''
+    echo 'Usage:'
+    echo '  raven_framework [-D][interfaceCheck][interactive][interactiveCheck] file'
+    echo ''
+    echo 'Description:'
+    echo '  Runs RAVEN using the input file given by "file".'
+    echo ''
+    echo 'Options:'
+    echo '  -D'
+    echo '    Development mode.  Turns Python "assert" statements on.'
```
Comment (wangcj05) [inline raven_framework]: add the description about the 'opt' mode

Comment (PaulTalbot-INL) [reply raven_framework]: I don't think so.  There is no flag or option for it, so it does not belong in the description of "help", I believe.

**Comment by wangcj05** [comment]
@PaulTalbot-INL In addition, an email should be proposed to the raven developers about these changes. Last comment, should we use 'xrange' instead of 'range' in the wiki optimization page?

File: raven_framework (hunk 5 above)
Comment (wangcj05) [reply raven_framework]: For me it is more clear to add this description. Because if the user just run 'raven_framework' without the input file, the current 'help' message will guide the user to run the code in 'dev' mode instead of 'opt' mode. I think we should put 'opt' there, and make it default as the code does.

**Comment by PaulTalbot-INL** [comment]
I do include a difference between iterators and lists, which is the difference between xrange and range, but if you want to specifically bring out that option, by all means add a section.

Proposed email:
We have recently modified how RAVEN runs to optimize performance.  In particular, we are now by default using the "-O" parameter in Python, which at this point simply removes assertion checks.  We only use assertions during development; they should not be reachable when running RAVEN as normal.

To summarize, you might see RAVEN speed up a little, but don't need to do or change anything if you're using RAVEN and not developing it.

For developers, passing the "-D" option to "raven_framework" will re-enable the "assert" statements, allowing for more effective debugging when implementing new features in RAVEN.

Thank you and enjoy!

**Comment by wangcj05** [comment]
The email is good to me, maybe you can add "For developers, we recommend to use 'assert' to do the internal check instead of 'if' statement."

**Comment by PaulTalbot-INL** [comment]
Added to the [development guide].(https://github.com/idaholab/raven/wiki/Developer_Information#assertions)

File: raven_framework (hunk 5 above)
Comment (wangcj05) [reply raven_framework]: any thoughts here?

Comment (PaulTalbot-INL) [reply raven_framework]: I don't feel like the help guides the user to run in dev mode, unless they feel like they want to run with assert statements on.  The usage syntax marks the default operation as being without any optional flags.  I'd rather remove the "help" than try and replicate the user manual in it; why do you feel the user is being guided towards dev mode?

Comment (PaulTalbot-INL) [reply raven_framework]: I increased the verbosity and added a "default usage" mode, which is somewhat unorthodox but might be more clear to users new to *nix.

Comment (wangcj05) [reply raven_framework]: First, I agree with you we do not want to replicate the user manual in the code. But a brief clear description I think is also useful. Second, I do not think we have documented this new feature in the user manual (we may not need now since this changes only affect the developers), that's why I suggest to put it here. Anyway, my thoughts may be wrong, I really appreciate that you made the changes. To me it is more clear, however it may the have no difference to the users.

File: raven_framework [hunk 6]
```diff
@@ -50,27 +50,36 @@ case $MODE in
...
+    echo '      Do not attempt to run the model executable, but read from existing output files.'
+    echo ''
+    echo '    interactive'
+    echo '      Run RAVEN in GUI mode if possible.'
+    echo ''
+    echo '    interactiveCheck'
+    echo '      Run tests on interactive GUI.'
     echo ''
     ;;
   *)
```
Comment (wangcj05) [inline raven_framework]: Thanks for the changes. To me it is more clear.

**Comment by joshua-cogliati-inl** [comment]
Just out of curiosity, when run_tests is done, does RAVEN run in debug or opt mode?

**Comment by wangcj05** [comment]
Thanks for the changes, and I think this PR is good to me.

**Comment by PaulTalbot-INL** [comment]
@joshua-cogliati-inl That's a great question.  I don't even know what the right answer is.

If you run tests in dev mode, and you fail an assertion, either 1) the test would have passed, which means the assertion is wrong and should probably be changed; or 2) the test couldn't have passed, in which case you're probably still developing, but maybe you still want to run tests with assertions present to see who fails.

If you run tests in opt mode, well, they will run microscopically faster.

This said, I guess I don't see much reason to run the tests in opt mode.  I'll see if I can change that easily.

**Comment by wangcj05** [comment]
It is good to me. What do you think? @joshua-cogliati-inl

**Comment by joshua-cogliati-inl** [comment]
"the test would have passed, which means the assertion is wrong and should probably be changed" or the assertion caught something that is a subtle problem that is not caught by actual test, as in I have had assertions fail that would result in a wrong number, but not an outright failure.

I don't have time to look at this, but I have no objections to it being merged.

**Comment by wangcj05** [comment]
OK, I will do the checklist, and additional issue #410 is created and will be resolved in the future PR.

**Comment by wangcj05** [comment]
Now this PR is ready to go. Thanks @PaulTalbot-INL, Great Job!
